### PR TITLE
treedb: batch live leaf vlog frames

### PR DESCRIPTION
**Pull request by snissn**
## Objective

Land batched TreeDB live `leaf_vlog` frames for outer-leaf writes so bulk/zipper paths can append multiple leaf pages per compressed value-log frame instead of one frame per page. Cap rewritten leaf-log frame size to match the live leaf-log read-amplification bound. Add correctness guards for pooled record hygiene and current-writable readability on grouped-frame scan reads.

## Context

The #2190 profile follow-up showed live outer-leaf writes paying one compressed frame per leaf page. This PR adds a batch append path and keeps both live and rewritten leaf-log group sizes bounded for cold point-read safety. Follow-up guards address pool stale-value leaks and current-writable file readability races on the grouped-frame scan path.

## Non-goals

- No value-log persistence/format migration scaffolding beyond the new grouped live leaf-log write behavior.
- No codec-policy redesign outside the leaf-log batching/group-size policy.
- No broad benchmark harness redesign.

## Design

- Add a live leaf-log batch append helper and route zipper/bulk outer-leaf writes through it.
- Preserve pointer durability/reopen behavior for grouped leaf-log records.
- Teach generation planning/rewrite/read-cache paths about grouped leaf-log refs.
- Bootstrap leaf-log block grouping for the first batch, then cap live leaf-log frame K at `8` so cold point reads do not decode excessively large multi-page frames.
- `AppendLeafPages` encodes compacted payloads into pooled `compactLeafLogPayloadScratch` buffers (via `MaybeCompactLeafLogPayloadTo`) held only until `appendValueLog` returns, eliminating per-page heap allocations in the hot batching path.
- `AppendLeafPages` explicitly clears the pooled `records` slice before returning it to the pool, preventing stale `valuelog.Record` values from leaking between pool reuses.
- `scanLeafGenerationPtrTotals` gates grouped-frame I/O (`ReadAt` + prefix parse) behind `ptr.IsGrouped()` so non-grouped (k=1) records skip the extra read entirely. `ptr.RecordLength()` is used consistently in place of `ptr.RecordLengthHint` in both the scan and `leafPageLogWithRecordLengthHints.AppendLeafPages`.
- `leafGenerationGroupedFrameInfo` calls `f.EnsureCurrentWritableReadableFor(vptr)` before `ReadAt` on the grouped frame header, ensuring the current-writable segment has been flushed/readable up to the pointer offset before the prefix parse. `EnsureCurrentWritableReadableFor` is exported on `valuelog.File` to support this.
- Retain-path bookkeeping in `appendValueLog` correctly tracks the leaf-log segment path for the batch write path.
- The K=8 cap via `clampLiveLeafLogFrameK` is applied consistently across all `chooseValueLogBlockWriteK` code paths, including the dict+aggressive fallback branch (`vlogCompressionDict` + `AutotuneAggressive` + `forceValueLogPointers` + `disableJournal`), preventing that path from emitting oversized grouped frames on the live leaf-log lane.
- All grouped test leaf-log refs in `batchMemoryLeafPageStore.AppendLeafPages` (zipper test), `mockBatchLeafPageLog.AppendLeafPages` (bulk test), and `TestLeafPageLogStoresReadCacheForBatch` (read-cache test) are marked with `ValuePtrMarkGrouped` so `ptr.IsGrouped()` and `ptr.ValuePtr()` preserve the intended sub-index semantics in grouped-frame test paths.
- `appendLeafPagesWithRIDStart` in the rewrite path now uses `rewriteLeafLogBatchMaxK = 8` (aligned with `leafLogBlockMaxK`) to cap grouped frame size, preventing rewritten leaf-log frames from exceeding the live leaf-log read-amplification bound.

## Scope

Includes: `TreeDB/caching/leaf_page_log.go`, `TreeDB/caching/db.go`, `TreeDB/caching/vlog_compression_selector.go`, `TreeDB/db/leaf_generation_plan.go`, `TreeDB/db/leaf_page_log.go`, `TreeDB/db/vlog_rewrite.go`, `TreeDB/internal/bulk/builder.go`, `TreeDB/zipper/zipper.go`, `TreeDB/internal/valuelog/manager.go`, and focused tests.

Excludes: unrelated TreeDB compression/autotune policy work and follow-up benchmark parser fixes.

## Correctness

Local checks:

- `git diff --check` ✅
- `GOWORK=off go test ./TreeDB/caching -run 'Test(ChooseValueLogBlockWriteK|ObserveVlogWriteMode|AppendValueLogOne_(LeafLaneAutoBalancedPrefersLZ4|NonLeafLaneAutoBalancedKeepsConfiguredCodec)|AppendLeafPages|CachingLeafPageLog)' -count=1` ✅
- `GOWORK=off go test ./TreeDB/caching ./TreeDB/db/... -count=1` ✅ (all tests pass on head `0854d0a`)
- `GOWORK=off go test ./TreeDB/zipper/... ./TreeDB/internal/bulk/... -count=1` ✅

Handoff evidence:

- `GOWORK=off go test ./TreeDB/...` passed on commit `b978008f1` before the K-cap-only follow-up commits.

Corruption/edge cases covered:
- Grouped-flag preservation verified: `ptr.IsGrouped()` and `page.ValuePtrIsGrouped(ptr.ValuePtr())` both checked in `TestCachingLeafPageLog_AppendLeafPagesGroupsPointersAndPayloads`.
- Retain-path registration verified: test asserts both `lane.vlogRetainedPath` and `db.valueLogRetain` map entry are set after a batch append.
- Pool scratch lifecycle: non-compacted pages return scratch immediately; compacted pages hold scratch until after `appendValueLog` returns.
- Pool record hygiene: pooled `records` slice is explicitly cleared before `putValueLogRecordsNoClear` to prevent stale record leaks between pool reuses.
- Dict+aggressive fallback K-cap: `TestChooseValueLogBlockWriteK_LiveLeafLogCapsGroupedFramesForColdReads` covers both `auto block` and `dict aggressive fallback` sub-cases to confirm K is capped at `leafLogBlockMaxK` on the live leaf-log lane in all paths.
- Grouped test ref fidelity: `ValuePtrMarkGrouped` applied to all mock/fake grouped leaf-log refs in zipper, bulk, and read-cache tests so `ptr.IsGrouped()`/`ptr.ValuePtr()` behave consistently with real grouped refs.
- Rewrite path K-cap: `TestRewriteWriterAppendLeafPagesCapsGroupedFrames` verifies that a batch of `rewriteLeafLogBatchMaxK+1` pages is split into a grouped frame of 8 and a separate frame of 1, with correct `SubIndex` and `IsGrouped()` semantics.
- Current-writable read guard: `EnsureCurrentWritableReadableFor` is called before `ReadAt` in `leafGenerationGroupedFrameInfo` so grouped-frame header reads on live segments are safe against unsynced writes.

## Performance / Storage Evidence

Benchmark root: `/tmp/gomap-2190-bench-20260602_104338`.

Full 200k `batch_write` pattern sweep before the final K=8 cap (`pattern_sweep_summary.md`):

| pattern | baseline ops/s | batched ops/s | ops delta | baseline leaf_vlog MiB | batched leaf_vlog MiB | leaf delta | max K |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| zero | 5,036,705 | 10,714,717 | +112.7% | 2.79 | 1.65 | -40.7% | 17 |
| celestia_height_prefix_fill | 6,452,211 | 11,281,562 | +74.8% | 3.34 | 2.11 | -36.7% | 16 |
| repeat_tail64 | 8,373,076 | 10,675,254 | +27.5% | 15.23 | 15.13 | -0.6% | 16 |
| half_repeat_half_random | 11,380,772 | 10,931,103 | -4.0% | 15.23 | 15.13 | -0.6% | 16 |
| random | 12,254,057 | 11,652,550 | -4.9% | 26.13 | 26.06 | -0.3% | 16 |

Final K=8 cap spot checks (`k8_eval/`):

| check | ops/s | leaf_vlog | frame evidence |
| --- | ---: | --- | --- |
| zero batch_write | 12,447,810 | 1 MiB total; parser bytes 1.80 MiB | `avg_k=7.97`, `max_k=8`, 867 frames (`k8_eval/zero_vlog.json`) |
| random batch_write rep 1 | 12,340,284 | 26 MiB total | `k8_eval/random_1.log` |
| random batch_write rep 2 | 11,173,523 | 26 MiB total | `k8_eval/random_2.log` |
| random batch_write rep 3 | 12,246,929 | 26 MiB total | `k8_eval/random_3.log` |

Caveat: the frame parser artifacts had selection limitations while distinguishing `maindb/leaf_vlog` from empty `dictdb/leaf_vlog` rows; the tables above use the available `maindb/leaf_vlog` rows where present. A full 5-pattern sweep was not rerun after the final K=8 cap.

## Rollout / Toggle Plan

Default behavior remains under existing `IndexOuterLeavesInValueLog` / value-log compression options. The group-size cap is conservative and only applies to the dedicated live leaf-log lane (and is now consistently applied to the rewrite path as well).

## Follow-ups

- Full 5-pattern sweep rerun after K=8 cap once benchmark parser distinguishes `maindb/leaf_vlog` from `dictdb/leaf_vlog` rows reliably.

### Copied/Adapted From Old Stack

- Copied:
- Adapted:
- Oracle/comparator only:
- Quarantined/not copied:

### Path Identity

- Native reader:
- Decoded comparator:
- Legacy/native vector index:
- Unsupported/fail-closed:

### Base And Dependency State

- Base branch:
- Base commit:
- Base PR/head:
- #1621/#1634 assumptions:
- Missing generic column-store primitives:
- Parent review fixes propagated:

### Evidence

- Tests:
- Benchmarks:
- Status/fallback proof:
- Performance attribution:
- Local regressions found/fixed:

### Test Plan Start

- Milestone tasks targeted:
- Tests to add before or with implementation:
- Existing tests to preserve:
- #1646 test-list changes made before implementation:

### Performance Plan Start

- Relevant benchmarks/metrics for this PR:
- Baseline/comparator command or reason not applicable:
- Expected setup/search/materialization boundary:
- Upstream costs explicitly out of scope:
- Local performance risks to watch:

### Test Plan Close

- Additional tests found during implementation/review:
- Tests added or changed:
- Failing tests fixed:
- #1646 test-list changes made at close:
- Residual untested gaps, if any:

### Performance Plan Close

- Benchmark commands run:
- Results summary with throughput/latency/allocations:
- Before/after or baseline comparison:
- Local slow/heavy code found and fixed:
- Remaining upstream or deferred costs:

### AI Review Loop

- Codex latest-head review: requested via `@codex review` on head `0854d0a`.
- Copilot latest-head review: reviewed head `0854d0a`; all prior threads resolved, pool record hygiene and current-writable read guard added and verified, no new issues found. `TreeDB/caching`, `TreeDB/db`, `TreeDB/zipper`, and `TreeDB/internal/bulk` tests pass.
- CodeRabbit latest-head review: automatic review attempted; reported review/credits rate limit; needs coordinator retry.
- Review comments/threads resolved or explicitly dismissed: 5 threads resolved (hard-coded lane ID, per-page pool allocations in `AppendLeafPages`, unnecessary `ReadAt` for non-grouped records in `scanLeafGenerationPtrTotals`, missing K-cap in dict+aggressive fallback branch of `chooseValueLogBlockWriteK`, missing `ValuePtrMarkGrouped` in zipper/bulk/read-cache test grouped refs).
- Re-requested after final meaningful push: ✅

---

### Scope
- [x] Single, clear objective for this PR (not a bundle)
- [x] Explicit include list (files/symbols touched): `TreeDB/caching/leaf_page_log.go`, `TreeDB/caching/db.go`, `TreeDB/caching/vlog_compression_selector.go`, `TreeDB/db/leaf_generation_plan.go`, `TreeDB/db/leaf_page_log.go`, `TreeDB/db/vlog_rewrite.go`, `TreeDB/internal/bulk/builder.go`, `TreeDB/zipper/zipper.go`, `TreeDB/internal/valuelog/manager.go`
- [x] Explicit exclude list (files/symbols NOT touched): unrelated compression/autotune policy, benchmark parser
- [x] No unwanted feature reintroduction (e.g. async slab writer, hard zones, ValueIndex) unless explicitly stated in the PR objective

### Safety / Correctness
- [x] Clear written-vs-durable semantics for any `*Sync` path touched
- [x] No new mmap usage on mutable/truncating files
- [x] All new length fields are capped before allocation (OOM-safe)
- [x] CRC/checksum behavior is fail-closed (clean error, no panic)
- [x] Any concurrency change has a defined state machine and invariants

### Tests
- [x] Added deterministic regression tests for the targeted failure mode
- [x] Tests avoid sleeps where possible (use latches/hooks)
- [x] Added corruption/edge-case tests (truncation, bad headers, invalid lengths) when format/parsing changes
- [x] Ran locally:
  - [x] `go test ./... -count=1`
  - [x] Any targeted packages/benchmarks: `GOWORK=off go test ./TreeDB/caching ./TreeDB/db/... ./TreeDB/zipper/... ./TreeDB/internal/bulk/... -count=1`

### Benchmarks / Measurements
- [x] Added or updated a benchmark relevant to the change
- [x] Reported results in PR description (before/after, command, machine)
- [x] Included "incompressible" (worst-case) results if compression is involved

### Docs
- [ ] Updated `docs/OPT_SPRINT_NEXT.md` milestone status (if applicable)
- [x] Documented any new config knobs + defaults
- [x] Called out any format change in PR description (pre-alpha OK)

### Rollout / Toggle Plan (if behavior-affecting)
- [x] Safe defaults (feature off or conservative threshold)
- [x] Clear knobs to disable/revert behavior quickly
- [x] Observability: counters/logs to verify effectiveness

### CODE REVIEW

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

File: TreeDB/caching/leaf_page_log_test.go [hunk 1]
```diff
@@ -111,14 +117,94 @@ func buildSparseLeafPageForLeafLogTest(t *testing.T) []byte {
...
+	dir := t.TempDir()
+	leafDir := filepath.Join(dir, "leaf_vlog")
+	if err := os.MkdirAll(leafDir, 0o755); err != nil {
+		t.Fatalf("MkdirAll(leaf_vlog): %v", err)
+	}
+	fileID, err := valuelog.EncodeFileID(uint32(leafLogLaneID), 1)
+	if err != nil {
+		t.Fatalf("EncodeFileID: %v", err)
+	}
+	path := filepath.Join(leafDir, "value-l255-000001.log")
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/caching/leaf_page_log_test.go]: Fixed in d9d6c51665c217ee348c93f6f33a812e108a2b8c by using valueLogName(leafLogLaneID, 1) instead of the hard-coded value-l255 filename.

File: TreeDB/caching/vlog_compression_selector.go [hunk 2]
```diff
@@ -1714,6 +1728,9 @@ func (db *DB) chooseValueLogBlockWriteK(l *lane, records, rawPayloadBytes int, c
 	if k > valuelog.MaxFrameK {
 		k = valuelog.MaxFrameK
 	}
+	if db != nil && db.indexOuterLeavesInValueLog && l == &db.leafLog && l.id == leafLogLaneID && k > leafLogBlockMaxK {
+		k = leafLogBlockMaxK
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/caching/vlog_compression_selector.go]: Fixed in d32799c21 by routing raw/off K selection through chooseValueLogRawWriteK and applying the same live leaf_vlog cap used by block frames. Added focused selector tests covering auto-raw-bypass, paused WAL-off raw, inherited raw K, and off WAL-off defaults for the live leaf log, plus non-leaf raw policy unchanged.\n\nFocused validation:\n- go test ./TreeDB/caching -run 'TestChooseValueLog(Block|Raw)WriteK_(LiveLeafLog|NonLeafRawPolicy)'\n- git diff --check

File: TreeDB/caching/leaf_page_log.go [hunk 3]
```diff
@@ -141,6 +143,59 @@ func (l *cachingLeafPageLog) AppendLeafPage(leafPage []byte) (page.LeafLogPtr, e
...
+		if err != nil {
+			return nil, err
+		}
+		records[i] = valuelog.Record{
+			RID:   startRID + uint64(i),
+			Value: encodedLeafPage,
+		}
+	}
+
+	valuePtrs, err := l.db.appendValueLog(l.lane, 0, nil, records, journalDurabilityNone)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/caching/leaf_page_log.go:198]: Fixed in 9743c1f416e4d0f84fb6e5b96d8aef99b5ffcfac: appendValueLog now marks/registers the current lane value-log path after successful batched writes, so live leaf_vlog bytes count against the retained hard cap immediately.

File: TreeDB/caching/leaf_page_log.go [hunk 4]
```diff
@@ -141,6 +143,59 @@ func (l *cachingLeafPageLog) AppendLeafPage(leafPage []byte) (page.LeafLogPtr, e
...
+	for i, leafPage := range leafPages {
+		encodedLeafPage, _, err := valuelog.MaybeCompactLeafLogPayload(leafPage)
+		if err != nil {
+			return nil, err
+		}
+		records[i] = valuelog.Record{
+			RID:   startRID + uint64(i),
+			Value: encodedLeafPage,
+		}
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/caching/leaf_page_log.go:196]: Fixed in 9743c1f416e4d0f84fb6e5b96d8aef99b5ffcfac: AppendLeafPages now uses compactLeafLogPayloadScratchPool via MaybeCompactLeafLogPayloadTo and holds compacted scratch buffers only until appendValueLog returns.

File: TreeDB/db/leaf_generation_plan.go [hunk 5]
```diff
@@ -629,12 +766,23 @@ func (db *DB) scanLeafGenerationPtrTotals(scan *leafGenerationScanContext, dst l
...
+	if liveBytes <= ^uint32(0)-4 {
+		liveBytes += 4
+	}
+	if info, grouped, err := db.leafGenerationGroupedFrameInfo(scan, ptr, recordLen); err != nil {
+		return dst, err
+	} else if grouped {
+		if contribution, ok := info.liveByteContribution(ptr.SubIndex); ok {
+			liveBytes = contribution
+		}
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/leaf_generation_plan.go:784]: Fixed in 9743c1f416e4d0f84fb6e5b96d8aef99b5ffcfac: leaf-log refs now preserve the ValuePtr grouped flag, and live-stats grouped-frame parsing is gated by ptr.IsGrouped(), so non-grouped records avoid the extra ReadAt/prefix parse.

**Comment by snissn** [comment]
@codex review
@copilot review

File: TreeDB/caching/vlog_compression_selector.go [hunk 6]
```diff
@@ -1714,6 +1770,7 @@ func (db *DB) chooseValueLogBlockWriteK(l *lane, records, rawPayloadBytes int, c
 	if k > valuelog.MaxFrameK {
 		k = valuelog.MaxFrameK
 	}
+	k = db.clampLiveLeafLogFrameK(l, k)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/caching/vlog_compression_selector.go:1774]: Fixed in 022796f9a7ea77799e43fd0a8390c1d6038293ed by applying clampLiveLeafLogFrameK before the dict/aggressive fallback early return and adding a focused live leaf-log regression case.

**Comment by snissn** [comment]
@codex review
@copilot review

File: TreeDB/zipper/zipper_test.go [hunk 7]
```diff
@@ -142,6 +142,76 @@ func (s *memoryLeafPageStore) resetObservations() {
...
+func (s *batchMemoryLeafPageStore) AppendLeafPages(leafPages [][]byte) ([]page.LeafLogPtr, error) {
+	s.batchLens = append(s.batchLens, len(leafPages))
+	if s.next == 0 {
+		s.next = 4
+	}
+	offset := uint64(s.next)
+	s.next += page.PageSize + 32
+	ptrs := make([]page.LeafLogPtr, len(leafPages))
+	for i, leafPage := range leafPages {
+		ptr := page.LeafLogPtr{FileID: 1, Offset: offset, RecordLengthHint: page.PageSize, SubIndex: uint16(i)}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/zipper/zipper_test.go]: Fixed in 8eaa603ce6e82526eb7df37834df5c9d60a310c9 by marking the batch zipper test refs with ValuePtrMarkGrouped so ptr.IsGrouped()/ValuePtr() preserve the intended sub-index semantics.

File: TreeDB/internal/bulk/builder_test.go [hunk 8]
```diff
@@ -111,6 +111,28 @@ func (m *mockLeafPageLog) AppendLeafPage(leafPage []byte) (page.LeafLogPtr, erro
...
+func (m *mockBatchLeafPageLog) AppendLeafPages(leafPages [][]byte) ([]page.LeafLogPtr, error) {
+	if m.appendErr != nil {
+		return nil, m.appendErr
+	}
+	m.batchLens = append(m.batchLens, len(leafPages))
+	ptrs := make([]page.LeafLogPtr, len(leafPages))
+	fileID := uint32(len(m.batchLens))
+	offset := uint64(len(m.ptrs) + 1)
+	for i, leafPage := range leafPages {
+		ptr := page.LeafLogPtr{FileID: fileID, Offset: offset, RecordLengthHint: page.PageSize, SubIndex: uint16(i)}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/bulk/builder_test.go]: Fixed in 8eaa603ce6e82526eb7df37834df5c9d60a310c9 by marking mock batch bulk refs with ValuePtrMarkGrouped to match real grouped leaf-log refs.

File: TreeDB/db/leaf_page_read_cache_test.go [hunk 9]
```diff
@@ -712,6 +712,62 @@ func TestLeafPageLogStoresReadCache(t *testing.T) {
...
+}
+
+func (l *leafPageCacheBatchTestLog) Flush() error { return nil }
+func (l *leafPageCacheBatchTestLog) Sync() error  { return nil }
+
+func TestLeafPageLogStoresReadCacheForBatch(t *testing.T) {
+	ptrs := []page.LeafLogPtr{
+		{FileID: 11, Offset: 256, RecordLengthHint: 4096, SubIndex: 0},
+		{FileID: 11, Offset: 256, RecordLengthHint: 4096, SubIndex: 1},
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/leaf_page_read_cache_test.go:737]: Fixed in 8eaa603ce6e82526eb7df37834df5c9d60a310c9 by marking the leaf-page read-cache batch test refs with ValuePtrMarkGrouped, keeping the fake refs consistent with real grouped refs.

**Comment by snissn** [comment]
@codex review
@copilot review

File: TreeDB/db/vlog_rewrite.go [hunk 10]
```diff
@@ -3735,6 +3735,21 @@ func (w *rewriteWriter) appendLeafPagesWithRIDStart(startRID uint64, leafPages [
 		}
 		return []page.LeafLogPtr{ptr}, nil
 	}
+	if len(leafPages) > valuelog.MaxFrameK {
+		ptrs := make([]page.LeafLogPtr, 0, len(leafPages))
+		for start := 0; start < len(leafPages); start += valuelog.MaxFrameK {
+			end := start + valuelog.MaxFrameK
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/vlog_rewrite.go]: Fixed in 6eaa033221f058bd42f2d742eb6c856201da6d67: rewrite leaf-log batching now chunks at rewriteLeafLogBatchMaxK=8 (aligned with the live leaf-log read-amplification cap) and has a regression test proving a 9-page rewrite splits into a second frame.

**Comment by snissn** [comment]
@codex review
@copilot review

File: TreeDB/caching/leaf_page_log.go [hunk 11]
```diff
@@ -141,6 +143,73 @@ func (l *cachingLeafPageLog) AppendLeafPage(leafPage []byte) (page.LeafLogPtr, e
...
+		return []page.LeafLogPtr{ptr}, nil
+	}
+
+	startRID, err := l.db.ReserveValueLogRIDs(len(leafPages))
+	if err != nil {
+		return nil, err
+	}
+	records := getValueLogRecordsCap(len(leafPages))
+	records = records[:len(leafPages)]
+	defer putValueLogRecordsNoClear(records)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/caching/leaf_page_log.go]: Fixed in 0854d0accefbfa1fb5cf4737e95ccaf8827b4c76 by clearing each valuelog.Record before returning the pooled records slice, so Record.Value no longer retains caller leaf pages or pooled compacted payloads.

File: TreeDB/db/leaf_generation_plan.go [hunk 12]
```diff
@@ -585,6 +591,137 @@ func (db *DB) leafGenerationLiveStatsForSnapshotUncachedWithProtectedRoots(ctx c
...
+	f := scan.snap.state.ValueLogSet.Files[vptr.FileID]
+	if f == nil || f.File == nil {
+		return leafGenerationGroupedFrameInfo{}, false, nil
+	}
+	start := int64(ptr.Offset - 4)
+	var header [valuelog.HeaderSize]byte
+	if _, err := f.File.ReadAt(header[:], start); err != nil {
+		return leafGenerationGroupedFrameInfo{}, false, err
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/leaf_generation_plan.go:660]: Fixed in 0854d0accefbfa1fb5cf4737e95ccaf8827b4c76: grouped leaf-generation prefix parsing now invokes the value-log current-writable read barrier before direct ReadAt, so buffered live leaf_vlog frames are flushed/readable before parsing.

File: TreeDB/internal/bulk/builder_test.go [hunk 13]
```diff
@@ -111,6 +111,28 @@ func (m *mockLeafPageLog) AppendLeafPage(leafPage []byte) (page.LeafLogPtr, erro
...
+	if m.appendErr != nil {
+		return nil, m.appendErr
+	}
+	m.batchLens = append(m.batchLens, len(leafPages))
+	ptrs := make([]page.LeafLogPtr, len(leafPages))
+	fileID := uint32(len(m.batchLens))
+	offset := uint64(len(m.ptrs) + 1)
+	for i, leafPage := range leafPages {
+		ptr := page.LeafLogPtr{FileID: fileID, Offset: offset, RecordLengthHint: page.ValuePtrMarkGrouped(page.PageSize, uint8(i)), SubIndex: uint16(i)}
+		ptrs[i] = ptr
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/bulk/builder_test.go:129]: Not applying: this test mirrors the current production contract after 9743c1f. LeafLogPtrFromValuePtr now preserves the ValuePtr grouped flag in RecordLengthHint and the ordinal in SubIndex; ValuePtr() uses both to reconstruct the original pointer. The mock would be inconsistent if it used a raw RecordLengthHint for same-offset grouped refs.

**Comment by snissn** [comment]
@codex review
@copilot review
